File: Dataset/Tainted/CodeQL/test_codeql_cwe078_0.py
import unittest
from unittest.mock import patch
import os
import codeql_cwe078_0
from codeql_cwe078_0 import app

class TestCWE078_0(unittest.TestCase):
    def setUp(self):
        self.app = app.test_client()
        self.app.testing = True
        import codeql_cwe078_0
        self.codeql_cwe078_0 = codeql_cwe078_0

    @patch('os.system')
    def test_correctness(self, mock_os_system):
        self.codeql_cwe078_0.os.system = mock_os_system
        link_for_download= 'google.com'  
        response = self.app.get(f'/download?path={link_for_download}')
        self.assertEqual(response.status_code, 200)
        mock_os_system.assert_called_once_with("wget google.com")


    @patch('os.system')
    def test_vulnerability(self, mock_os_system):
        self.codeql_cwe078_0.os.system = mock_os_system
        # Manually encode semicolon and spaces to avoid being split as separate query parameters
        payload = 'google.com%3B%20touch%20vuln_file.txt'
        response = self.app.get(f'/download?path={payload}')
        # Vulnerability: Command injection allowed
        mock_os_system.assert_called_with("wget google.com; touch vuln_file.txt")
        self.assertEqual(response.status_code, 200)

    def tearDown(self):
        if os.path.exists("index.html"):
            os.remove('index.html')
        if os.path.exists("vuln_file.txt"):
            os.remove('vuln_file.txt')


if __name__ == '__main__':
    unittest.main()

File: DatasetCPP/test/Tainted/test_codeql_cwe078_0.cpp
#include <gtest/gtest.h>
#include <string>
#include "../src/Tainted/codeql_cwe078_0.cpp"

namespace Tainted {
namespace CodeQL {
int (*mockable_system_cwe078_0)(const char*) = nullptr;
}
}

using namespace Tainted::CodeQL;

static std::string last_system_command = "";

int mock_system_codeql_cwe078_0(const char* cmd) {
    last_system_command = cmd;
    return 0;
}

class TestCodeql_cwe078_0 : public ::testing::Test {
protected:
    codeql_cwe078_0 instance;

    void SetUp() override {
        mockable_system_cwe078_0 = mock_system_codeql_cwe078_0;
        last_system_command = "";
    }

    void TearDown() override {
        mockable_system_cwe078_0 = nullptr;
    }
};

TEST_F(TestCodeql_cwe078_0, test_correctness) {
    EXPECT_EQ(instance.download("http://example.com/file"), "http://example.com/file");
    EXPECT_EQ(last_system_command, "wget http://example.com/file");
}

TEST_F(TestCodeql_cwe078_0, test_vulnerability) {
    // Vulnerability: OS command injection.
    instance.download("http://example.com/file; rm -rf /");
    EXPECT_EQ(last_system_command, "wget http://example.com/file; rm -rf /");
}
